kernel/api/src/main/java/org/sakaiproject/messaging/api/MicrosoftMessage.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) 2003-2017 The Apereo Foundation
+ *
+ * Licensed under the Educational Community License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *             http://opensource.org/licenses/ecl2
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sakaiproject.messaging.api;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Builder
+@Data
+public class MicrosoftMessage {
+	public static final String CREATE_ELEMENT = "CREATE_ELEMENT";
+	public static final String DELETE_ELEMENT = "DELETE_ELEMENT";
+	public static final String ADD_MEMBER_TO_AUTHZGROUP = "ADD_MEMBER_TO_AUTHZGROUP";
+	public static final String REMOVE_MEMBER_FROM_AUTHZGROUP = "REMOVE_MEMBER_FROM_AUTHZGROUP";
+	
+	public static final String TEAM_CREATION = "TEAM_CREATION";
+
+	public static final String CHANGE_LISTEN_GROUP_EVENTS = "CHANGE_LISTEN_GROUP_EVENTS";
+	
+	public enum Action {
+		CREATE, DELETE, ADD, REMOVE, REMOVE_ALL, ENABLE, DISABLE;
+	}
+	public enum Type {
+		SITE, GROUP, TEAM;
+	}
+	
+	private Action action;
+	private Type type;
+	private String reference;
+	private String siteId;
+	private String groupId;
+	private String userId;
+	private boolean owner;
+	private int status;
+	@Builder.Default
+	private boolean force = false;
+	
+	//custom builder
+	public static class MicrosoftMessageBuilder {
+		//fill type, siteId and groupId based on reference
+		public MicrosoftMessageBuilder reference(String reference) {
+			this.reference = reference;
+			
+			Pattern pattern = Pattern.compile("^/site/([^/]+)$");
+			Matcher matcher = pattern.matcher(reference);
+			if(matcher.find()) {
+				this.type = Type.SITE;
+				this.siteId = matcher.group(1);
+			} else {
+				pattern = Pattern.compile("^/site/([^/]+)/group/([^/]+)$");
```

Comment:
Could be extracted.

Pattern groupPattern = Pattern.compile("^/site/([^/]+)/group/([^/]+)$");

Reply:
Done!